### PR TITLE
refactor: arm

### PR DESCRIPTION
**Pull request by ProgrammingSR**

### CODE REVIEW

**Comment by LeonardAbbas** [comment]
worked on the arm today. It can be driven manually using the driver controller. Code stops are in place to prevent the motor from extending beyond a reasonable limit. It drives very smoothly. It needs a dynamic height and extension stop for the robot to be legal. Code to drive the arm to a position is written but not tested.

**Comment by ReeledWarrior14** [comment]
I would like to test this tomorrow before merging.

**Review by ArnoUUU — DISMISSED** [review]
I'm assuming new constants and other values are accurate.

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -65,6 +66,16 @@ public class RobotContainer {
   public RobotContainer() {
     configureButtonBindings();
 
+    m_armSubsystem.setDefaultCommand(
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/RobotContainer.java:69]: Operator joystick controls should go under driver controls for formatting purposes

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -145,43 +141,18 @@ private void configureButtonBindings() {
     new JoystickButton(m_operatorController, Button.kA.value)
         .onTrue(new InstantCommand(grabberSubsystem::toggle, grabberSubsystem));
 
+    // Deploys arm when left bumper is held and retracts when released
+    new JoystickButton(m_operatorController, Button.kLeftBumper.value)
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/RobotContainer.java:145]: Is there a functional purpose to this?

File: src/main/java/frc/robot/subsystems/ArmSubsystem.java [hunk 3]
```diff
@@ -10,365 +10,101 @@
...
-    }
-    
-    
-    //Check for overextentsion, of so decrease the setpoint
-    if (m_elevatorPID.getSetpoint() > ArmConstants.kMaxSwitchPos - 0.01) {
-      m_elevatorPID.setSetpoint(ArmConstants.kMaxSwitchPos - 0.01);
-      SmartDashboard.putString("return", "overext");
-      return;
+    // Prevents pivot from extending beyond reasonable limits
+    if (m_pivotEncoder.getAbsolutePosition() < ArmConstants.kPivotMinPosition && m_pivotSpeed < 0) {
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/subsystems/ArmSubsystem.java:54]: Should be less than or equal to

File: src/main/java/frc/robot/subsystems/ArmSubsystem.java [hunk 4]
```diff
@@ -10,365 +10,101 @@
...
+      m_pivotSpeed = 0;
+    } else if (m_pivotEncoder.getAbsolutePosition() > ArmConstants.kPivotMaxPosition && m_pivotSpeed > 0) {
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/subsystems/ArmSubsystem.java:56]: Should be greater than or equal to

File: src/main/java/frc/robot/subsystems/ArmSubsystem.java [hunk 5]
```diff
@@ -10,365 +10,101 @@
...
+      m_pivotSpeed = 0;
     }
 
-    //Don't allow retraction before seeing limit switch
-    if (m_elevatorPID.getSetpoint() < getElevatorEncoder() && !seenSwitch) {
-      m_elevatorPID.setSetpoint(getElevatorEncoder());
-      SmartDashboard.putString("return", "limitR");
-      return;
+    // Prevents elevator from extending until it hits the stop
+    if (m_elevatorMotor.getEncoder().getPosition() < ArmConstants.kElevatorMinPosition && m_elevatorSpeed < 0) {
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/subsystems/ArmSubsystem.java:61]: Should be less than or equal to

File: src/main/java/frc/robot/subsystems/ArmSubsystem.java [hunk 6]
```diff
@@ -10,365 +10,101 @@
...
+      m_elevatorSpeed = 0;
+    } else if (m_elevatorMotor.getEncoder().getPosition() > ArmConstants.kElevatorMaxPosition && m_elevatorSpeed > 0) {
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/subsystems/ArmSubsystem.java:63]: Should be greater than or equal to

File: src/main/java/frc/robot/subsystems/ArmSubsystem.java [hunk 7]
```diff
@@ -10,365 +10,101 @@
...
-      double motorSpeed = MathUtil.clamp(es, -0.25, 0.25)
-      + (0.015 * Math.cos(m_pivotEncoder.getAbsolutePosition()));
-      m_elevatorMotor.set(motorSpeed);
-      SmartDashboard.putNumber("motorSpeed", motorSpeed);
-    } else {
-      SmartDashboard.putNumber("motorSpeed", 999);
+    // If arm is above legal height, bring it down
+    double heightOverLimit = m_elevatorMotor.getEncoder().getPosition()
+        - (ArmConstants.kMaxGameHeight / Math.cos(Math.toRadians(m_pivotEncoder.getAbsolutePosition())));
+    if (heightOverLimit > 0 && m_elevatorSpeed > -heightOverLimit * 2) {
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/subsystems/ArmSubsystem.java:70]: Should be "if heightOverLimit is greater than or equal to 0"

File: src/main/java/frc/robot/subsystems/ArmSubsystem.java [hunk 8]
```diff
@@ -10,365 +10,101 @@
...
+      m_elevatorSpeed = -heightOverLimit * 2;
     }
 
+    // Counters the effect of gravity with a small feed forward
+    m_pivotMotor.set(m_pivotSpeed - (Math.sin(Math.toRadians(m_pivotEncoder.getAbsolutePosition())) * 0.03));
```
Comment (ReeledWarrior14) [inline src/main/java/frc/robot/subsystems/ArmSubsystem.java:75]: Feed forward speeds should be in constants

**Review by ReeledWarrior14 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotContainer.java (hunk 2 above)
Comment (LeonardAbbas) [reply src/main/java/frc/robot/RobotContainer.java:145]: We will need more buttons for other positions but having the arm auto retract so it isn't extended while driving is useful.

**Comment by LeonardAbbas** [comment]
tested and works with maximum height and maximum extension. We believe the elevator conversion constant from rotations to meters is slightly off, making us a little bit off. It will need to be updated. Also the positions to pick up cones from will need still need to be tuned.

**Review by ArnoUUU — DISMISSED** [review]
Tested

File: src/main/java/frc/robot/Robot.java [hunk 9]
```diff
@@ -38,12 +37,12 @@ public void robotPeriodic() {
 
   @Override
   public void disabledInit() {
```
Comment (mebrahimaleem) [inline src/main/java/frc/robot/Robot.java:39]: Robot.disabledInit() function should be removed entirely

Comment (ReeledWarrior14) [reply src/main/java/frc/robot/Robot.java:39]: Why? There is no reason to remove it. It is a default function added by WPILib.

**Review by ArnoUUU — APPROVED** [review]

**Review by ReeledWarrior14 — APPROVED** [review]

**Pull request merged**